Add support for Okta oauth2 integration

maheshlokhande1 opened the pull request:
- [x] Issue: #160
- [x] List of related PRs. : None
- [x] Project builds without any errors.
- [x] Unit Tests updated (or created where needed) and all pass.
- [x] You have ran manual functional “smoke” test (does product as a whole still work?).
- [x] User Story meets the acceptance criteria - and passes. acceptance criteria should be understood at outset.
- [x] There is a PR for update of internal documentation for the affected Project(s) - README etc.
- [x] Description of the changes: #160 
- [x] Signed-off-by: Mahesh Lokhande <mahesh.lokhande@ibm.com>
[Okta_settings.docx](https://github.com/IBM/lmctl/files/9888092/Okta_settings.docx)

dvaccarosenna commented:
@maheshlokhande1 all PRs should be targeting the `develop` branch. Would you be ok to update this PR?

maheshlokhande1 commented:
> @maheshlokhande1 all PRs should be targeting the `develop` branch. Would you be ok to update this PR?

Done.

dvaccarosenna commented on lmctl/cli/commands/login.py:
```diff
@@ -25,34 +24,37 @@
         ('client_id', '--client-id'),
         ('client_secret', '--client-secret'),
         ('auth_address', '--auth-address'),
+        ('auth-server-id', '--auth_server_id'),
```
`scope` and `is_okta` will also need to be mutually exclusive with `token`

maheshlokhande1 replied:
Addressed

dvaccarosenna commented on lmctl/client/api/auth.py:
```diff
@@ -42,6 +44,41 @@ def request_user_access(self, client_id: str, client_secret: str, username: str,
+        auth_response = self.base_client.make_request_for_json(request)
+        return auth_response
+
+    def request_okta_access(self, client_id: str, client_secret: str, scope: str = None, auth_server_id: str = None, okta_server: str = None) -> Dict:
```
This should probably be `request_okta_client_access`?

maheshlokhande1 replied:
Addressed

dvaccarosenna commented on lmctl/drivers/lm/security.py:
```diff
@@ -76,7 +79,14 @@ def __init__(self, auth_address, username=None, password=None, client_id=None, c
                 if self.__client_id is not None:
-                    client_builder.user_pass_auth(username=self.__username, password=self.__password, client_id=self.__client_id, client_secret=self.__client_secret)
+                    if self.__auth_mode == OKTA_MODE:
+                        client_builder.okta_user_pass_auth(username=self.__username, password=self.__password,
```
What about `okta_client_credentials_auth`? That could be used? I think trying to combine the checks for Okta with the legacy Oauth has led to this confusion. 
Can we add `elif self.__auth_mode.lower() == OKTA_AUTH_MODE:` under the check for Zen/Token and handle Okta separate? Leaving the `else` block for `#Oauth` as it was?

maheshlokhande1 replied:
Addressed

dvaccarosenna commented on lmctl/environment/lmenv.py:
```diff
@@ -87,6 +93,17 @@ def _validate_oauth(cls, values):
 
+    @classmethod
+    def _validate_okta(cls, values):
+        cls._validate_oauth(values)
```
Errors thrown by `_validate_oauth` will mention oauth mode not okta mode, which will be confusing:

```
Secure TNCO environment must be configured with either "client_id" or "username" property when using "auth_mode=oauth}"
```

Might be best to write all validation logic specific to Okta in `_validate_okta` rather than rely on the existing oauth logic (even though it's similar)

maheshlokhande1 replied:
Addressed

dvaccarosenna reviewed (changes requested):
Comments added. I also have some general comments:

- Unit tests need to be added for Okta (you will find where we added similar tests for Zen)
- Documentation needs to be updated for Okta. 
  - [Getting Started](https://github.com/IBM/lmctl/blob/master/docs/getting-started.md#oauth-authentication) should include steps for logging in with Okta
  - [Command reference for the login command](https://github.com/IBM/lmctl/blob/master/docs/command-reference/login.md) should be updated with the new parameters
  - [Example config in the configuration guide](https://github.com/IBM/lmctl/blob/master/docs/configure.md) should include an example of Okta
  - Check for other places that Okta specific docs may be needed

maheshlokhande1 commented:
> Comments added. I also have some general comments:
> 
> * Unit tests need to be added for Okta (you will find where we added similar tests for Zen)
> * Documentation needs to be updated for Okta.
>   
>   * [Getting Started](https://github.com/IBM/lmctl/blob/master/docs/getting-started.md#oauth-authentication) should include steps for logging in with Okta
>   * [Command reference for the login command](https://github.com/IBM/lmctl/blob/master/docs/command-reference/login.md) should be updated with the new parameters
>   * [Example config in the configuration guide](https://github.com/IBM/lmctl/blob/master/docs/configure.md) should include an example of Okta
>   * Check for other places that Okta specific docs may be needed

Updated docs.

dvaccarosenna commented on docs/configure.md:
```diff
@@ -209,6 +209,59 @@ environments:
+
+      #====================================#
+      # Oauth Option 1:                    #
+      # (Unoffical)TNCO Client credentials #
```
This says TNCO client credentials but it's actually Okta from my understanding? Also is the `(Unofficial)` part true?

dvaccarosenna commented on docs/getting-started.md:
````diff
@@ -174,6 +174,35 @@ If the output of this command ends with `CP4NA orchestration tests passed! ✅`
+Run `lmctl login`, changing the values to valid parameters for your environment:
+
+```
+lmctl login $API_GATEWAY --okta --client <Client> --client-secret <secret> -u <user> -p <password> --scope <scope>  --auth-address https://<okta server endpoint> --auth-server-id <auth_server_id>```
````
Can we have a follow-up story which gives an example of how to obtain the relevant argument values from Okta? I think we'll have a lot of developers/testers in need of that kind of information

maheshlokhande1 replied:
Yes we can create another story, I have already attached doc for Okta configuration in the issue.

dvaccarosenna commented on lmctl/drivers/lm/security.py:
```diff
@@ -71,12 +74,25 @@ def __init__(self, auth_address, username=None, password=None, client_id=None, c
+            if self.__username is not None:
+                # Using password auth
+                if self.__client_id is not None:
+                    client_builder.okta_user_pass_auth(username=self.__username, password=self.__password,
```
If `self.__username is not None` but `self.__client_id is None` then nothing happens - should this be considered an error scenario?

maheshlokhande1 replied:
Addressed.

dvaccarosenna replied:
I think the error needs to be clearer: `TNCO environment cannot be configured without a client_id when using auth_mode={OKTA_MODE}`

maheshlokhande1 replied:
Addressed

dvaccarosenna commented on lmctl/environment/lmenv.py:
```diff
@@ -87,6 +93,23 @@ def _validate_oauth(cls, values):
+        if not values.get('auth_server_id', None):
+            raise ValueError(f'Secure TNCO environment must be configured with "auth_server_id" when using "auth_mode={OKTA_MODE}". If the TNCO environment is not secure then set "secure" to False')
+        if values.get('username', None) and not values.get('scope', None):
+            raise ValueError(f'Secure TNCO environment cannot be configured with "scope" when using "auth_mode={OKTA_MODE}". If the TNCO environment is not secure then set "secure" to False')
```
This says `Secure TNCO environment cannot be configured with "scope" when using "auth_mode={OKTA_MODE}"` but don't we actually want to say that scope must be set when using Okta mode with a username?

maheshlokhande1 replied:
Addressed.

dvaccarosenna commented on tests/integration/integration_test_properties.py:
```diff
@@ -10,25 +10,29 @@ def __init__(self, client_credentials: ClientCredentialsAuth,
             legacy_user_pass=LegacyUserPassAuth(**legacy_user_pass) if legacy_user_pass is not None else None,
-            token_auth=JwtTokenAuth(**token_auth) if token_auth is not None else None
+            token_auth=JwtTokenAuth(**token_auth) if token_auth is not None else None,
+            okta_user_pass_auth=OktaUserPassAuth(**okta_user_pass_auth)
```
Should this include ` if okta_user_pass_auth is not None else None`? Like the other lines?

maheshlokhande1 replied:
Addressed.

maheshlokhande1 commented on lmctl/client/api/auth.py:
```diff
@@ -33,7 +35,7 @@ def request_user_access(self, client_id: str, client_secret: str, username: str,
             'username': username,
             'password': password,
-            'grant_type': 'password'
+            'grant_type': 'password',
```
Addressed.

maheshlokhande1 commented on docs/configure.md:
```diff
@@ -209,6 +209,59 @@ environments:
+
+      #====================================#
+      # Oauth Option 1:                    #
+      # (Unoffical)Okta Client credentials #
```
Addressed.

dvaccarosenna replied:
What is the outcome of this? If it's unofficial because CP4NA does not support it then let's remove it as an example

maheshlokhande1 replied:
Addressed.
